# Optimizing tensorboard

In [1]:
from IPython.display import YouTubeVideo
#Load video 5/11 from the below playlist URL
YouTubeVideo('https://www.youtube.com/watch?list=PLQVvvaa0QuDfhTox0AjmQ6tvTgMBZBEXN&v=j-3vuBynnOE', width=1000, height=500)

In [2]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Flatten, Dropout, Activation, Dense, Conv2D, MaxPooling2D
import pickle
import time
from tensorflow.keras.callbacks import TensorBoard

tf.__version__

'1.13.1'

In [3]:
X = pickle.load(open('../Loading Own dataset/X.pickle', 'rb'))
y = pickle.load(open('../Loading Own dataset/y.pickle', 'rb'))

In [4]:
X = X/255.0

### Previously

In [5]:
NAME = "Optimizing Tensorboard {}".format(int(time.time()))

In [6]:
tensorboard = TensorBoard(log_dir='logs/{}'.format(NAME))

In [7]:
model = Sequential()

model.add(Conv2D(64, (3,3), input_shape = X.shape[1:]))
model.add(Activation('elu'))
model.add(MaxPooling2D(pool_size=(2,2)))

model.add(Conv2D(64, (3,3)))
model.add(Activation('elu'))
model.add(MaxPooling2D(pool_size=(2,2)))

model.add(Flatten())
model.add(Dense(1))
model.add(Activation('sigmoid'))

model.compile(loss = 'binary_crossentropy',
             optimizer = 'adam',
             metrics = ['accuracy'])

model.fit(X, y, batch_size = 32, epochs = 10, validation_split = 0.1, callbacks = [tensorboard])

Instructions for updating:
Colocations handled automatically by placer.
Train on 22451 samples, validate on 2495 samples
Instructions for updating:
Use tf.cast instead.
Epoch 1/10
22451/22451 [==============================] - 12s 549us/sample - loss: 0.5944 - acc: 0.6796 - val_loss: 0.5349 - val_acc: 0.7359
Epoch 2/10
22451/22451 [==============================] - 8s 359us/sample - loss: 0.5225 - acc: 0.7455 - val_loss: 0.5152 - val_acc: 0.7387
Epoch 3/10
22451/22451 [==============================] - 9s 401us/sample - loss: 0.4900 - acc: 0.7666 - val_loss: 0.5018 - val_acc: 0.7531
Epoch 4/10
22451/22451 [==============================] - 8s 365us/sample - loss: 0.4680 - acc: 0.7814 - val_loss: 0.4975 - val_acc: 0.7587
Epoch 5/10
22451/22451 [==============================] - 8s 372us/sample - loss: 0.4387 - acc: 0.7984 - val_loss: 0.4994 - val_acc: 0.7623
Epoch 6/10
22451/22451 [==============================] - 10s 447us/sample - loss: 0.4105 - acc: 0.8123 - val_loss: 0.5011 - val_a

### For Optimization

In [8]:
dense_layers = [0, 1, 2]
layer_sizes = [32, 64, 128]
conv_layers = [1, 2, 3]

In [14]:
for dense_layer in dense_layers:
    for layer_size in layer_sizes:
        for conv_layer in conv_layers:
            Name = "{}-conv {}-nodes {}-dense {}-time".format(conv_layer, layer_size, dense_layer, int(time.time()))
            tensorboard = TensorBoard(log_dir='Optimization_logs/{}'.format(Name))
            print(Name)
            Optimized_model = Sequential()
            
            Optimized_model.add(Conv2D(layer_size, (3,3), input_shape = X.shape[1:]))
            Optimized_model.add(Activation('elu'))
            Optimized_model.add(MaxPooling2D(pool_size= (2,2)))
            
            for size in range(conv_layer-1):
                Optimized_model.add(Conv2D(layer_size, (3,3)))
                Optimized_model.add(Activation('elu'))
                Optimized_model.add(MaxPooling2D(pool_size = (2,2)))
                
            Optimized_model.add(Flatten())
            
            for dense in range(dense_layer):
                Optimized_model.add(Dense(layer_size))
                Optimized_model.add(Activation('elu'))
                
            Optimized_model.add(Dense(1))
            Optimized_model.add(Activation('sigmoid'))
            
            Optimized_model.compile(loss='binary_crossentropy',
                       optimizer = 'adam',
                       metrics = ['accuracy'])

            Optimized_model.fit(X, y, batch_size = 32, epochs = 10, validation_split = 0.1, callbacks=[tensorboard])

1-conv 32-nodes 0-dense 1556727683-time
Train on 22451 samples, validate on 2495 samples
Epoch 1/10
22451/22451 [==============================] - 5s 207us/sample - loss: 0.6439 - acc: 0.6277 - val_loss: 0.6046 - val_acc: 0.6645
Epoch 2/10
22451/22451 [==============================] - 4s 179us/sample - loss: 0.5948 - acc: 0.6842 - val_loss: 0.5786 - val_acc: 0.6998
Epoch 3/10
22451/22451 [==============================] - 4s 178us/sample - loss: 0.5620 - acc: 0.7162 - val_loss: 0.5631 - val_acc: 0.7086
Epoch 4/10
22451/22451 [==============================] - 4s 184us/sample - loss: 0.5396 - acc: 0.7337 - val_loss: 0.5600 - val_acc: 0.7174
Epoch 5/10
22451/22451 [==============================] - 4s 188us/sample - loss: 0.5213 - acc: 0.7477 - val_loss: 0.5418 - val_acc: 0.7166
Epoch 6/10
22451/22451 [==============================] - 4s 179us/sample - loss: 0.5080 - acc: 0.7535 - val_loss: 0.5556 - val_acc: 0.7194
Epoch 7/10
22451/22451 [==============================] - 4s 187us/samp

22451/22451 [==============================] - 9s 392us/sample - loss: 0.3417 - acc: 0.8496 - val_loss: 0.4614 - val_acc: 0.7840
Epoch 6/10
22451/22451 [==============================] - 9s 403us/sample - loss: 0.2978 - acc: 0.8735 - val_loss: 0.4543 - val_acc: 0.7988
Epoch 7/10
22451/22451 [==============================] - 9s 397us/sample - loss: 0.2514 - acc: 0.8946 - val_loss: 0.4855 - val_acc: 0.7936
Epoch 8/10
22451/22451 [==============================] - 10s 425us/sample - loss: 0.2117 - acc: 0.9116 - val_loss: 0.5426 - val_acc: 0.7796
Epoch 9/10
22451/22451 [==============================] - 10s 452us/sample - loss: 0.1734 - acc: 0.9318 - val_loss: 0.6295 - val_acc: 0.7772
Epoch 10/10
22451/22451 [==============================] - 10s 433us/sample - loss: 0.1393 - acc: 0.9473 - val_loss: 0.6546 - val_acc: 0.7772
1-conv 128-nodes 0-dense 1556728172-time
Train on 22451 samples, validate on 2495 samples
Epoch 1/10
22451/22451 [==============================] - 11s 496us/sample - 

22451/22451 [==============================] - 6s 252us/sample - loss: 0.2032 - acc: 0.9179 - val_loss: 0.6044 - val_acc: 0.7743
3-conv 32-nodes 1-dense 1556728906-time
Train on 22451 samples, validate on 2495 samples
Epoch 1/10
22451/22451 [==============================] - 6s 261us/sample - loss: 0.6126 - acc: 0.6546 - val_loss: 0.5205 - val_acc: 0.7383
Epoch 2/10
22451/22451 [==============================] - 5s 223us/sample - loss: 0.5117 - acc: 0.7454 - val_loss: 0.4897 - val_acc: 0.7651
Epoch 3/10
22451/22451 [==============================] - 5s 227us/sample - loss: 0.4598 - acc: 0.7791 - val_loss: 0.4633 - val_acc: 0.7743
Epoch 4/10
22451/22451 [==============================] - 5s 241us/sample - loss: 0.4215 - acc: 0.8050 - val_loss: 0.4381 - val_acc: 0.7960
Epoch 5/10
22451/22451 [==============================] - 5s 234us/sample - loss: 0.3893 - acc: 0.8223 - val_loss: 0.4478 - val_acc: 0.7968
Epoch 6/10
22451/22451 [==============================] - 5s 234us/sample - loss: 

22451/22451 [==============================] - 22s 963us/sample - loss: 0.4405 - acc: 0.7920 - val_loss: 0.5026 - val_acc: 0.7511
Epoch 5/10
22451/22451 [==============================] - 19s 854us/sample - loss: 0.3856 - acc: 0.8237 - val_loss: 0.5253 - val_acc: 0.7455
Epoch 6/10
22451/22451 [==============================] - 21s 932us/sample - loss: 0.2946 - acc: 0.8705 - val_loss: 0.5352 - val_acc: 0.7651
Epoch 7/10
22451/22451 [==============================] - 19s 866us/sample - loss: 0.1765 - acc: 0.9306 - val_loss: 0.6833 - val_acc: 0.7459
Epoch 8/10
22451/22451 [==============================] - 20s 911us/sample - loss: 0.0888 - acc: 0.9668 - val_loss: 0.9175 - val_acc: 0.7467
Epoch 9/10
22451/22451 [==============================] - 19s 861us/sample - loss: 0.0657 - acc: 0.9763 - val_loss: 1.3270 - val_acc: 0.7623
Epoch 10/10
22451/22451 [==============================] - 22s 977us/sample - loss: 0.0437 - acc: 0.9847 - val_loss: 1.0996 - val_acc: 0.7655
3-conv 128-nodes 1-dens

22451/22451 [==============================] - 8s 351us/sample - loss: 0.0569 - acc: 0.9796 - val_loss: 1.3632 - val_acc: 0.7106
Epoch 10/10
22451/22451 [==============================] - 8s 373us/sample - loss: 0.0474 - acc: 0.9842 - val_loss: 1.5248 - val_acc: 0.7046
2-conv 64-nodes 2-dense 1556730288-time
Train on 22451 samples, validate on 2495 samples
Epoch 1/10
22451/22451 [==============================] - 8s 350us/sample - loss: 0.5931 - acc: 0.6726 - val_loss: 0.5446 - val_acc: 0.7246
Epoch 2/10
22451/22451 [==============================] - 7s 312us/sample - loss: 0.4976 - acc: 0.7569 - val_loss: 0.5027 - val_acc: 0.7507
Epoch 3/10
22451/22451 [==============================] - 7s 325us/sample - loss: 0.4264 - acc: 0.8008 - val_loss: 0.5110 - val_acc: 0.7579
Epoch 4/10
22451/22451 [==============================] - 7s 327us/sample - loss: 0.3365 - acc: 0.8524 - val_loss: 0.5172 - val_acc: 0.7659
Epoch 5/10
22451/22451 [==============================] - 8s 334us/sample - loss:

## Result

2-conv 64-nodes 2-dense 1556730288-time performed the best with an 

    accuracy of 0.9879
    Loss     of 0.03494

In [15]:
model.save('64x3-CNN.model')